Feature/spectral error indicator

adperezm opened the pull request:
Added a module to calculate the spectral error indicator.

Some parts of the code are lifted from @adampep implementation. I abused "associate" on purpose to make it easier for him to modify everything when he needs to extend this for AMR.

Most of the work here were additions to neko that do not affect the rest of componets.

To the type(coef_t) I added the transformation matrices needed to go to Legendre space. This are needed for compression, spectral error indicators, and some filtering if someone wants it later. It did not make sense to implement them separately. Currently the cpr module has the functions as well but I will update it that independently to use coef.

timofeymukha commented:
Is the deletion of rayleigh.f90 intentionally part of this one?

adperezm commented:
> 

Most certainly not! I made a mistake. Thanks for catching that!

MartinKarp reviewed (commented):
I think a things I am struck by is whether this should be in coef, to me it seems to be more related to the space and independent of the geometry. Maybe the mapping should be in space instead?

adperezm commented:
Yeap, seeing the space type... It should 100% be there.

timofeymukha commented:
Jeez Im really bad at using this review tool...

timofeymukha commented on src/sem/spectral_error_indicator.f90:
```diff
@@ -0,0 +1,676 @@
+                
+    integer :: i, j, k, e, nxyz, nelv, n
+    character(len=LOG_SIZE) :: log_buf 
+    character(len=4) :: space 
```
This was about space, since it is a dummy argument

adperezm commented:
> Jeez Im really bad at using this review tool...

I am very happy it is not me :)

timofeymukha commented:
I am actually thinking that the legendre could be its own small type, why not? Our base SEM classes are huge already. And if we want we can add the legendre class as a component to the space anyway. 

Then my usual complaint is the naming for this stuff.  Can we please be more verbose, *especially* if this will not be in its own type?
```fortran
     !> Legendre transformation matrices
     real(kind=rp), allocatable :: v(:,:)        !< legendre to physical
     real(kind=rp), allocatable :: vt(:,:)       !< legendre to physical t
     real(kind=rp), allocatable :: vinv(:,:)     !< Physical to legendre
     real(kind=rp), allocatable :: vinvt(:,:)    !< Physical to legendre t
     !> Legendre weights in matrix form
     real(kind=rp), allocatable :: w(:,:)        !< Legendre weights
```
Suggestion
```fortran
to_physical
to_physical_t
to_legendre
to_legendre_t
legendre_weights.
```` 
If this every gets too big we have `associate`, and in fact you use it already :-).

adperezm commented on src/sem/spectral_error_indicator.f90:
```diff
@@ -0,0 +1,676 @@
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+!
+!> Compression
```
Well, I have clearly been copying and pasting things :)

adperezm commented on src/sem/spectral_error_indicator.f90:
```diff
@@ -0,0 +1,676 @@
+  !! @param coef type with all geometrical variables
+  subroutine spec_err_ind_init(this, u_,v_,w_,coef)
+    class(spec_err_ind_t), intent(inout) :: this
+    type(field_t), intent(in), target :: u_
```
I will do so. It was there because before I had another attribute called "v" :) So I had to change naming conventions for a bit

adperezm commented:
So I will put the required thing in space, not create a new type. (There are conflicting request between you, and I will take the executive decision :) )

And I will change the namings. Which I knew from the begining I should have done :)

adperezm commented on src/sem/spectral_error_indicator.f90:
```diff
@@ -0,0 +1,676 @@
+     !> Working field - Consider making this a simple array
+     type(field_t) :: wk
+     !> Configuration of spectral error calculation
+     real(kind=rp) :: SERI_SMALL
```
I understand but I will not change it for the simple fact that I will let Adam do it when he gets into this :)

adperezm commented:
In the interest of pushing code and not getting lost, I am reopening this.

I still need to move the transformation stuff to space so it is not ready for merging

adperezm commented:
I have seen that space changed since last night so I will stop.

timofeymukha commented:
> I have seen that space changed since last night so I will stop. A warning would have been appreciated.

Yeah, it's the stuff Niclas warned about on zulip, that a bunch of main types including space_t will now have TBPs, as pretty much the rest of the code already has. Not too much to be scared of here, basically instead of stuff like `space_init(space,....)` you have `space%init(....)`.

adperezm commented:
> Adalberto, you are on a roll! But how about those variable names for the transforms, as I commented before? Do you think we could switch to something more verbose?
> 
> I think also the docstring for the spectral_error_indicator type needs some love. Also, your indentation there with `contains` is off, just like in the other PR.
> 
> Right, also in the constructor for the indicator, call the destructor as the first thing you do inside it. This is a convention we have in many types. The reason is that in Fortran nothing prevents one running init on an object has has already be meddled with somehow, so this ensures we make it a clean slate first.

Thanks!, I used the same file as a template for this module so that is the reason the same wrong stuff is in both. 

I fixed indentation now and added docstrings etc. I also added a %free call at the beginning.

The only thing I still have not changed is the naming of the transformation matrices. I still feel that they should be called with a conventional name. I used "V" because I saw it in some reference and Nek5000 uses it as well. I have also seen "T".

I am up for maybe adding some _to physical or something but I am not sure it should only be called that. 

I do see that one would need to specify from which space one goes and to which one the transformation takes you.

This because the case could come when one somehow goes to fourier space or something like that, and then the "to_physical" would be a different matrix than the "to_physical" from legendre.

timofeymukha commented:
> This because the case could come when one somehow goes to fourier space or something like that, and then the "to_physical" would be a different matrix than the "to_physical" from legendre.

Which would be totally solved by putting said matrices inside a small type, kekeke :-). But don't worry, we can always do that later if needed and new transforms arrive.

timofeymukha commented on src/sem/space.f90:
```diff
@@ -444,5 +514,99 @@ subroutine space_compute_dist(dx, x, lx)
+  
+  
+  !> Generate spectral tranform matrices
+  !! @param coef type with all geometrical variables
```
```suggestion
  !! @param Xh SEM function space.
```

adperezm replied:
Thanks! this was a left over of when this routine was in coef

timofeymukha commented on src/sem/spectral_error_indicator.f90:
```diff
@@ -0,0 +1,699 @@
+  !! @param u field to transform
+  !! @param wk working field
+  !! @param coef type coef for mesh parameters
+  !! @param space string that indicates which space to transform
```
```suggestion
  !! @param space String that indicates which space to transform, "spec" or "phys".
```

timofeymukha commented:
@adperezm, just added two fixes for docstrings, I think one was just incorrect (coef instead of space) and then for the other one I added the possible options for the parameter. Just commit them if you agree. Otherwise, green light from my side!

We will obviously go through this type again when GPU porting is done, which may offer a chance to work a bit with the legacy CPU code here as well.

adperezm commented:
> @adperezm, just added two fixes for docstrings, I think one was just incorrect (coef instead of space) and then for the other one I added the possible options for the parameter. Just commit them if you agree. Otherwise, green light from my side!
> 
> We will obviously go through this type again when GPU porting is done, which may offer a chance to work a bit with the legacy CPU code here as well.

Thank you very much!

Yes, I was hoping Adam would do it but it might be that I will implement it in GPU so we can see how to write those parts better when I understand them
